### PR TITLE
Node.js: Add node 8 and npm 5 support

### PR DESCRIPTION
**Pull request by ovaar**

### CODE REVIEW

**Comment by malikolivier** [comment]
Thanks for the update! Can you explain why you exclude npm from the build image in ca0417b1c77a8aebe05e95cf40bf55ea01114a1b?

This causes a QA error during package as the node_modules files as `/usr/lib/node_modules` is not included `FILES`.

What's more, let me just give my two cents. On a freshly installed ubuntu 16.04.3 (no GUI), I had to install a few packages for `do_compile` to succeed:
```sh
dpkg --add-architecture i386 && apt-get update
apt-get install lib32z1-dev libssl-dev:i386 gcc-multilib g++-multilib
```
I don't understand myself why I needed i386 headers, but I would get a linker error without them.

**Comment by ovaar** [comment]
[ca0417b](https://github.com/imyller/meta-nodejs/commit/ca0417b1c77a8aebe05e95cf40bf55ea01114a1b) was for personal interest. I did not realize it would append to the pull request. I'll revert that commit.

**Comment by ovaar** [comment]
Can we finalize this pull request?

**Comment by aaronovz1** [comment]
I second that^

File: classes/npm-base.bbclass [hunk 1]
```diff
@@ -1,7 +1,7 @@
 DEPENDS += " node-native"
 
 inherit nodejs-arch
-
+
```
Comment (bachp) [inline classes/npm-base.bbclass:4]: Unnecessery whitespace?

File: classes/npm-base.bbclass [hunk 2]
```diff
@@ -51,15 +59,19 @@ oe_runnpm() {
 
 	bbnote ${NPM} --registry=${NPM_REGISTRY} ${ARCH_FLAGS} ${NPM_FLAGS} "$@"
 
-	export JOBS=${@oe.utils.cpu_count()}
-
+	export JOBS="${@oe.utils.cpu_count()}"
+
```
Comment (bachp) [inline classes/npm-base.bbclass]: Unnecessery whitespace?

File: classes/npm-base.bbclass [hunk 3]
```diff
@@ -95,15 +107,19 @@ oe_runnpm_native() {
...
 	export NPM_CONFIG_DEV="false"
 
+	bbnote NPM version: $NPM_VERSION
 	bbnote NPM native architecture: ${NPM_ARCH_NATIVE}
 	bbnote NPM home directory: ${NPM_HOME_DIR_NATIVE}
 	bbnote NPM cache directory: ${NPM_CONFIG_CACHE}
 	bbnote NPM registry: ${NPM_REGISTRY}
 	bbnote NPM workdir .npmignore: ${NPM_IGNORE}
 
+
```
Comment (bachp) [inline classes/npm-base.bbclass]: Unnecessery newline?

File: classes/npm-base.bbclass [hunk 4]
```diff
@@ -114,8 +130,12 @@ oe_runnpm_native() {
...
+	if [ "${@npm_get_version()[0]}" = "5" ]; then
+		NPM_CACHE_CMD="verify"
+	fi
+
+	${NPM_NATIVE} cache $NPM_CACHE_CMD || die "oe_runnpm_native failed (cache clean)"
 
 	LD="${NPM_LD_NATIVE}" ${NPM_NATIVE} --registry=${NPM_REGISTRY} ${ARCH_FLAGS} ${NPM_FLAGS_NATIVE} "$@" || die "oe_runnpm_native failed (install)"
 
-}
+}
```
Comment (bachp) [inline classes/npm-base.bbclass]: Don't remove trailing newline

**Review by bachp — COMMENTED** [review]
I added some style related comments

File: classes/npm-base.bbclass [hunk 5]
```diff
@@ -79,10 +90,10 @@ NPM_FLAGS_NATIVE_append_class-nativesdk = " --unsafe-perm"
...
 
 	if [ "${NPM_ARCH_NATIVE}" != "allarch" ]; then
-                ARCH_FLAGS="--arch=${NPM_ARCH_NATIVE} --target_arch=${NPM_ARCH_NATIVE}"
-        else
-                ARCH_FLAGS=""
-        fi
+        ARCH_FLAGS="--arch=${NPM_ARCH_NATIVE} --target_arch=${NPM_ARCH_NATIVE}"
+    else
+        ARCH_FLAGS=""
+    fi
```
Comment (bachp) [inline classes/npm-base.bbclass]: Indention inconsistent, please align the if too.

**Review by bachp — CHANGES_REQUESTED** [review]
There are still some whitepsace and style issue. Please fix, the rest is fine.

**Comment by ovaar** [comment]
all good?

**Review by bachp — APPROVED** [review]
All good from my side

**Comment by mbbauer** [comment]
Could we please merge this?  The lts version of node has changed.

**Comment by ovaar** [comment]
@imyller  We need you man!

**Comment by mitorn** [comment]
FYI: `os.popen` is deprecated (used in `npm-base.bbclass`) in python 2.6 and `subprocess` module should be used instead. Also, `npm_get_version` in `npm-base.bbclass` leaves file open which will produce warning during build.

I fixed this as below patch suggests:
```
diff --git a/classes/npm-base.bbclass b/classes/npm-base.bbclass
index b545225..b40f241 100644
--- a/classes/npm-base.bbclass
+++ b/classes/npm-base.bbclass
@@ -19,10 +19,6 @@ NPM_FLAGS ?= ""
 
 NPM_FLAGS_append_class-nativesdk = " --unsafe-perm"
 
-def npm_get_version():
-    import os
-    return os.popen("npm --v").read().split('.')
-
 # Target npm
 
 oe_runnpm() {
@@ -66,7 +62,8 @@ oe_runnpm() {
 
        export HOME="${NPM_HOME_DIR}"
 
-       if [ "${@npm_get_version()[0]}" = "5" ]; then
+
+       if [ "$(echo ${NPM_VERSION} | cut -d. -f1)" = "5" ]; then
                NPM_CACHE_CMD="verify"
        fi
 
@@ -128,7 +125,7 @@ oe_runnpm_native() {
 
        export HOME="${NPM_HOME_DIR_NATIVE}"
 
-       if [ "${@npm_get_version()[0]}" = "5" ]; then
+       if [ "$(echo ${NPM_VERSION} | cut -d. -f1)" = "5" ]; then
                NPM_CACHE_CMD="verify"
        fi
 

```

**Comment by uwburn** [comment]
Any news on pulling this in? Having the new LTS would be great!

**Comment by zonque** [comment]
Is this repository still actively maintained or should people be using some replacement?

**Comment by aaronovz1** [comment]
Doesn't seem to be very active apart from thirdparty contributors.

**Comment by zonque** [comment]
Which is okay. I'm just wondering whether this is still the official nodejs layer repo, given that this PR is pending since 8 month.

**Comment by aaronovz1** [comment]
I maintain a fairly up to date PR and fork of this repo. Other than that the author of this repo has gone quiet for some time.

**Comment by ovaar** [comment]
+1

**Comment by ovaar** [comment]
@zonque there is no 'official' nodejs layer. This is the one though. Perhaps the author can authorization some other participants for maintaining this repository.

**Comment by mitorn** [comment]
There is some support for nodejs in openembedded-core. npm class seems to be quite much the same as in here. Also, devtool is supporting npm packages at least for some extent. I am investigating currently if i could start using devtool and openembedded-core npm class.

**Comment by deinok** [comment]
Guys, should we contact Node.js Foundation in order to create an oficial layer?

**Comment by keeslinp** [comment]
@mitorn Did your investigations have any success?

**Comment by bachp** [comment]
@mitorn The `npm.bbclass` in core use the bitbake fetcher to fetch all dependency. To use it `devtool` is pretty much a must as otherwise the dependency management is way to much work.

The classes in this layer are much simpler and they rely on `npm` to fetch everything. They do not use the bitbake fetcher mechanism and thus you can't cache/archive the downloads the way you normally do with bitbake. An "advantage" is that you do not need to specify all the dependency at the bitbake level. This might be easier in some cases especially when used with a local npm mirror.

@deinok Is there a need for a seperate layer. The latest LTS (8.x) version is also available in [meta-openembedded](http://cgit.openembedded.org/meta-openembedded/tree/meta-oe/recipes-devtools/nodejs). If somebody would be willing to maintain it I think it would also be possible to have the Latest (10.x) available as a recipe too.

**Comment by mitorn** [comment]
@keeslinp we haven't had time lately to test devtool based npm fetcher. It would be quite feasible for us since it does bitbake license management instead of creating your own. There was open questions at least about locking commits (can we trust package-lock-.json for example).

@bachp we would like to use `devtool` because of the license handling which i think this layer does not provide.

**Pull request closed without merging**